Add Judoc syntax reference (#1934)
anoma/juvix: docs/reference/judoc.md

# Judoc syntax reference
Judoc is used to document parts of your code. You can attach *Judoc
blocks* to the following entities:

1. A module.
2. A type definition.

> 4. A constructor definition.
> 3. A type signature of a function.
> 4. An axiom definition.

Reviewer: ```suggestion
3. A constructor definition.
4. A type signature of a function.
5. An axiom definition.
```

> In order to attach documentation to either of these items, write *blocks* of

Reviewer: You call the items in the list entities in the sentence above, and either usually refers to two alternatives.

```suggestion
In order to attach documentation to any of these entities, write *blocks* of
```

```suggestion
In order to attach documentation to an entity, write *blocks* of
```

documentation before them:

1. For modules:
   ```
   --- This module is cool
   module Cool;
   ..
   ```
2. For type definitions:
   ```
   --- Unary representation of natural numbers
   type Nat : Type :=
     | --- Nullary constructor representing number 0
       zero : Nat;
     | --- Unary constructor representing the successor of a natural number
       suc : Nat -> Nat;
   ```
3. For type signatures (and likewise for axioms):
   ```
   --- The polymorphic identity function
   id : {A : Type} -> A -> A;
   ```
   
   Next we define the syntax of Judoc *blocks*.

## Block

A *block* can be one of these:
1. A *paragraph*.
2. An *example*.

*Blocks* are separated by a line with only `---`.
For instance, this is a sequence of two *blocks*:
```
--- First block
---
--- Second block
```

Note that the following is a single block since it lacks the `---` separator:
```

> --- First block
> 
> 
> --- Still first block

Reviewer: Suggestion: Have 1 blank line between the two block to make it clear that only 1 and not 2 blank lines are required.

```suggestion
--- First block

--- Still first block
```

```

### Paragraph
A *paragraph* is a non-empty sequence of *lines*.

> For instance, the following is a paragraph with two *lines*:

Reviewer: Does the following actually have one line?

Author: I've fixed that, thanks

# Judoc syntax reference
Judoc is used to document parts of your code. You can attach *Judoc
blocks* to the following entities:

1. A module.
2. A type definition.
3. A constructor definition.
4. A type signature of a function.
5. An axiom definition.

In order to attach documentation to any of these entities, write _blocks_ of
documentation before them:

1. For modules:
   ```
   --- This module is cool
   module Cool;
   ..
   ```
2. For type definitions:
   ```
   --- Unary representation of natural numbers
   type Nat : Type :=

>      | --- Nullary constructor representing number 0
>        zero : Nat;

Reviewer: Remove trailing `;`

```suggestion
     | --- Nullary constructor representing number 0
       zero : Nat
```
